### ISSUE TITLE
Kiln thinks that Three Parts Date has an error

### ISSUE THREAD

**Issue by BryceStevenWilley** [plocket/docassemble-cucumber#704]
## Background

The [ALAffidavitOfIndigency](https://github.com/SuffolkLITLab/docassemble-ALAffidavitOfIndigency) uses the ThreePartDate. When running tests for the MotionToStayEviction, which uses the Affidavit, I ran into the following error:

```
The test tried to continue to the next page after qualify-on-face, but the user got an error. Check the picture of this page.
```

However, the screenshot wasn't of that screen (which only has a continue button) at all; it was of the next screen, asking the user to fill in their birthdate, a value that is also in the story table.

Trying to figure out where the error is actually through, I print debugged until I realized that `scope.waitUntilContinued` calls `scope.checkForError` before actually continuing, and `checkForError` was finding an error on the page. Couldn't really figure out much more about the error due to the [fact that ElementHandlers just print out `JSHandle@node`](https://github.com/puppeteer/puppeteer/issues/4976) when turned into strings. So I had to extend the debug window's timeout a bit to search for the same query selectors that we do in `checkForError`, and found that the ThreePartDate has an element that has the `da-has-error` class, it's just blank.

```html
<div class="col-md-8 dafieldpart">
    <div class="al_three_parts_date form-row row">
        <div class="col col-month">
            <label for="_ignore_dXNlcnNbMF0uYmlydGhkYXRl_month">Month</label>
            <select class="form-select al_field month dXNlcnNbMF0uYmlydGhkYXRl" id="dXNlcnNbMF0uYmlydGhkYXRl_month" name="_ignore_dXNlcnNbMF0uYmlydGhkYXRl_month" aria-describedby="dXNlcnNbMF0uYmlydGhkYXRl-error" required="required">
                <option value=""></option>
                <option value="01">January</option>
                <option value="02">February</option>
                <option value="03">March</option>
                <option value="04">April</option>
                <option value="05">May</option>
                <option value="06">June</option>
                <option value="07">July</option>
                <option value="08">August</option>
                <option value="09">September</option>
                <option value="10">October</option>
                <option value="11">November</option>
                <option value="12">December</option>
            </select>
        </div>
        <div class="col col-3 col-day">
            <label for="_ignore_dXNlcnNbMF0uYmlydGhkYXRl_day">Day</label>
            <input
                class="form-control al_field day dXNlcnNbMF0uYmlydGhkYXRl"
                type="number"
                min="1"
                inputmode="numeric"
                id="dXNlcnNbMF0uYmlydGhkYXRl_day"
                name="_ignore_dXNlcnNbMF0uYmlydGhkYXRl_day"
                aria-describedby="dXNlcnNbMF0uYmlydGhkYXRl-error"
                required="required"
            />
        </div>
        <div class="col col-3 col-year">
            <label for="_ignore_dXNlcnNbMF0uYmlydGhkYXRl_year">Year</label>
            <input
                class="form-control al_field year dXNlcnNbMF0uYmlydGhkYXRl"
                type="number"
                min="1"
                inputmode="numeric"
                id="dXNlcnNbMF0uYmlydGhkYXRl_year"
                name="_ignore_dXNlcnNbMF0uYmlydGhkYXRl_year"
                aria-describedby="dXNlcnNbMF0uYmlydGhkYXRl-error"
                required="required"
            />
        </div>
        <div id="al_dXNlcnNbMF0uYmlydGhkYXRl_error" class="da-has-error al_error"></div>
    </div>
    <input alt="Input box" class="form-control daBirthDate" type="hidden" name="dXNlcnNbMF0uYmlydGhkYXRl" id="dXNlcnNbMF0uYmlydGhkYXRl" required="" aria-hidden="true" style="display: none;" />
</div>
```

## Potential fixes:

* ignore things found by the search that have no inner text and not child elements. Seems hacky, but it works.
* fix ALToolbox to not have a div with the class `da-has-error`. This would involve more jquery stuff, and I'm not sure how all the validator works. So while it's a bit cleaner, I'm not sure how to approach it.

**Comment by plocket** [comment]
From the PR:

> It's true "fixing" ThreePartsDate would be better. It's possible to put a totally different class on the container of the errors, but then we do have to make sure that the errors themselves do get that class and there was some kind of barrier to that at the time. The code has changed a lot, though. There's a way with the jQuery validation plugin to give the errors themselves a custom class instead of using the default class. I think if we lose the default class we may have to change the css and bootstrap may not recognize our invalid input class. We'll have to experiment when there's bandwidth.

**Comment by BryceStevenWilley** [comment]
> There's a way with the jQuery validation plugin to give the errors themselves a custom class instead of using the default class. 

To clarify, the problem isn't with the default class; it's with the fact that the element has the class, even when there are no error (upon immediately loading the page). The user doesn't notice it because the element doesn't have any content. I was proposing that we find a way to only add the error class when there is an error on the page, as opposed to always having the error class, and filling it's element with content. I agree that changing that class would be difficult on the bootstrap side.

**Comment by plocket** [comment]
(More previous comments are in the PR, which we may close eventually.) There's at least a partial fix in AssemblyLine now. Before closing this, we need to write a test to do the following:
1. Enter invalid input
2. Enter new valid input
3. Try to continue

This is because there's a `display: none` `.da-has-error` that's around after any invalid input.

**Comment by BryceStevenWilley** [comment]
There are more issues with Three Parts Date now: https://github.com/SuffolkLITLab/docassemble-ALAffidavitOfIndigency/actions/runs/6087576039/job/16516429502. I'm not sure if it's this same issue, but Kiln doesn't log enough information to find out what's happening, and I don't have time to dive deep into why we can't fill in the Date.

**Comment by nonprofittechy** [comment]
Possibly an error showing a similar bug: https://github.com/SuffolkLITLab/docassemble-MA209AProtectiveOrder/actions/runs/6301668025

**Comment by plocket** [comment]
> There are more issues with Three Parts Date now

I believe this is a different issue. We had to disable some tests recently for other reasons and it looks like we never updated our internal tests to isolate tests for AL custom dates separately.

What I believe is the source of the bug: There was recently a bug where interacting with a hidden (show if, I believe) date field blocked the field just below. We then started skipping hidden fields to avoid the bug. Unfortunately, the way we interact with the ThreePartsDate/BirthDate fields is through their hidden field because of complications in trying to get it to otherwise submit the final value.

We'll have to find another solution to one of those problems.

Edit: see #764
